doc/developer/design/20230411_envelope_upsert_order_by_timestamp.md

```diff
@@ -0,0 +1,92 @@
+- Feature name: Envelope upsert with order by timestamp (limited scope)
+- Associated: https://github.com/MaterializeInc/materialize/issues/16512
+
+# Summary
+
+- Support user specified order by in Kafka upsert sources in a limited capacity. If users want to specify custom ordering which is semantically meaningful to them, there's currently no easy way to do that. They could potentially use `ENVELOPE NONE` along with `DISTINCT ON` to get similar results, but it's cumbersome and lacks the automatic compaction on keys with normal `ENVELOPE UPSERT`.
+- Currently, will limit the scope to allow only `ORDER BY` the `TIMESTAMP` metadata field behind an unsafe flag. Depending upon user feedback this can later be expanded to support expressions and extractions from other columns like `HEADERS`.
+
+# Scope
+
+The scope will be limited to only allow `ORDER BY TIMESTAMP` (default ordering `ASC`).
+
+- For kafka upsert style topics, value = NULL is taken to mean that it’s a delete tombstone and that the corresponding key should be removed. If we want to keep the same behavior it would make sense to limit the `ORDER BY` columns to only the metadata columns.
+- Limiting it to only `TIMESTAMP` because
+    - `OFFSET` is already default. We can make it explicit though if we want
+    - `PARTITION`, `KEY`, `TOPIC`, would not make a good order by clause
+    - `HEADERS` by itself would not also make a good order by clause (it’s a json blob), but later we can probably allow expressions so that values could be extracted from it.
+- Limiting to only `ASC` as default, because for compaction it makes sense to take semantically the _latest_ value of something, which implies an `ASC` order by
```

Comment:
Are you talking about kafka compaction here? I think you can simplify it to this:

```suggestion
- Limiting to only ascending order because no users have expressed a need for descending.
```

Reply:
I was referring to the `upsert` itself (since we are sort of doing the compaction on the key space), but I think this update reads better.